### PR TITLE
SDK/CLI: improve mask support in the auto-annotation functionality

### PR DESCRIPTION
**Pull request by SpecLad**
### Motivation and context

While it is already possible to output mask shapes from AA functions, which the driver _will_ accept, it's not convenient to do so. Improve the practicalities of it in several ways:

* Add `mask` and `polygon` helpers to the interface module.

* Add a helper function to encode masks into the format CVAT expects.

* Add a built-in torchvision-based instance segmentation function.

* Add an equivalent of the `conv_mask_to_poly` parameter for Nuclio functions.

### How has this been tested?

Manual and unit tests.

### Checklist

- [x] I submit my changes into the `develop` branch
- [ ] I have created a changelog fragment 
- [x] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- ~~[ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))~~
- ~~[ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/cvat-ai/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/cvat-ai/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/cvat-ai/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/cvat-ai/cvat/tree/develop/cvat-ui#versioning))~~

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

### CODE REVIEW

File: cvat-sdk/cvat_sdk/auto_annotation/functions/torchvision_instance_segmentation.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+        if context.conv_mask_to_poly:
+            # Since we treat mask values of exactly LEVEL as true, we'd like them
+            # to also be considered high by find_contours. And for that, the level
+            # parameter must be slightly less than LEVEL.
+            contours = measure.find_contours(
+                mask[0].detach().numpy(), level=math.nextafter(LEVEL, 0)
+            )
+            if not contours:
+                return []
```
Comment (zhiltsov-max) [inline cvat-sdk/cvat_sdk/auto_annotation/functions/torchvision_instance_segmentation.py:48]: 1. Aren't the masks binary here? Typically, masks are either binary or indexed, where the index is the instance class id or number
1. A single mask can consist of several polygons, so it makes sense to support this. Currently I can see only the first contour is used
1. It seems that it extracts contours >= level as negative- and the opposite as positive-oriented, so maybe you need to use only the negative ones to avoid producing duplicated annotations for holes, we had such problem in the original datumaro implementation

File: tests/python/sdk/test_masks.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        # 1100
+
+        assert encode_mask(bitmap, bbox) == [1, 5, 2, 1, 0, 4, 1]
+
+        # Edge case: full image
+        bbox = [0, 0, 6, 2]
+        assert encode_mask(bitmap, bbox) == [2, 3, 2, 2, 3, 0, 0, 5, 1]
+
+    def test_encode_mask_invalid(self):
+        with pytest.raises(ValueError):
```
Comment (zhiltsov-max) [inline tests/python/sdk/test_masks.py:51]: Consider using a regex to match the error message text

File: tests/python/sdk/test_masks.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        # 0111
+        # 1100
+
+        assert encode_mask(bitmap, bbox) == [1, 5, 2, 1, 0, 4, 1]
+
+        # Edge case: full image
+        bbox = [0, 0, 6, 2]
+        assert encode_mask(bitmap, bbox) == [2, 3, 2, 2, 3, 0, 0, 5, 1]
+
+    def test_encode_mask_invalid(self):
```
Comment (zhiltsov-max) [inline tests/python/sdk/test_masks.py:50]: Consider splitting the test into 3 separate ones

File: cvat-sdk/cvat_sdk/masks.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+# SPDX-License-Identifier: MIT
+
+import math
+from collections.abc import Sequence
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+
+def encode_mask(bitmap: ArrayLike, /, bbox: Sequence[float]) -> list[float]:
```
Comment (zhiltsov-max) [inline cvat-sdk/cvat_sdk/masks.py:12]: Why the bbox is required?

Comment (SpecLad) [reply cvat-sdk/cvat_sdk/masks.py:12]: Mainly because I thought I could get away with it. 🙂 Torchvision provides a bounding box with each mask for free, and so does Ultralytics, so I figured it was common enough that I didn't need to implement automatic bbox determination.

If you think it's warranted, I could do it, though.

File: tests/python/sdk/test_masks.py (hunk 2 above)
Comment (SpecLad) [reply tests/python/sdk/test_masks.py]: Sure, done.

File: cvat-sdk/cvat_sdk/auto_annotation/functions/torchvision_instance_segmentation.py (hunk 1 above)
Comment (SpecLad) [reply cvat-sdk/cvat_sdk/auto_annotation/functions/torchvision_instance_segmentation.py]: > Aren't the masks binary here?

I'm not sure which masks you're talking about, but `mask` is a float tensor. The torchvision recommends converting it to boolean by comparing it to 0.5, which is what I do in the non-polygon branch.

> A single mask can consist of several polygons, so it makes sense to support this.

Yeah, you're right. I originally tried this and I thought `find_contours` returned duplicate contours sometimes, which is why I only used the first contour. But I can't reproduce this problem anymore, so I must've made some mistake before. 🤷‍♂️ 

I changed the logic back to process all contours.

> maybe you need to use only the negative ones to avoid producing duplicated annotations for holes

Yep, implemented.

File: cvat-sdk/cvat_sdk/auto_annotation/functions/torchvision_instance_segmentation.py [hunk 5]
```diff
@@ -15,6 +17,38 @@
...
+        # parameter must be slightly less than LEVEL.
+        contours = measure.find_contours(mask[0].detach().numpy(), level=math.nextafter(LEVEL, 0))
+
+        for contour in contours:
+            if len(contour) < 3 or _is_positively_oriented(contour):
+                continue
+
+            contour = measure.approximate_polygon(contour, tolerance=2.5)
+
+            yield cvataa.polygon(label.item(), contour[:, ::-1].ravel().tolist())
```
Comment (zhiltsov-max) [inline cvat-sdk/cvat_sdk/auto_annotation/functions/torchvision_instance_segmentation.py:46]: Probably, these polygons should be grouped, if there are more than 1 of them.

Comment (zhiltsov-max) [reply cvat-sdk/cvat_sdk/auto_annotation/functions/torchvision_instance_segmentation.py:46]: As an option, assign a group, equal to the index, to each annotation, if no group is assigned already.

Comment (SpecLad) [reply cvat-sdk/cvat_sdk/auto_annotation/functions/torchvision_instance_segmentation.py:46]: I probably didn't describe the problem clearly enough: the problem is selecting group IDs that do not exist in the **remote** dataset. The user has the option to add the auto-generated annotations to existing annotations in the task, and it's important that new shapes are not added to existing groups.

As far as I can see, your suggestions only make sure that the auto-generated group IDs don't conflict with each other, which is good, but not sufficient.

I could solve this problem by fetching existing annotations, finding the maximum existing group ID and assigning new group IDs starting from that, but I don't really want to do that. It's very inefficient, and is susceptible to race conditions.

IMO, a proper solution lies beyond the scope of this PR and would involve server-side code. (E.g. the server could support special group values that would mean "assign a new group".)

Comment (zhiltsov-max) [reply cvat-sdk/cvat_sdk/auto_annotation/functions/torchvision_instance_segmentation.py:46]: Ok, I thought the annotations are supposed to be overwritten in the result of the call. If not, then it indeed requires a more complicated solution.

Comment (SpecLad) [reply cvat-sdk/cvat_sdk/auto_annotation/functions/torchvision_instance_segmentation.py:46]: The overwrite is optional (it's controlled by the `clear_existing` flag). This mirrors the "Clean previous annotations" checkbox in the UI.

**Review by zhiltsov-max — APPROVED** [review]
